draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -816,327 +1115,323 @@ Input: alpha, an octet string to be hashed.
+2.   t1 = Z * u^2
+3.   t2 = t1^2
+4.   x1 = t1 + t2
+5.   x1 = inv0(x1)
+6.   e1 = x1 == 0
+7.   x1 = x1 + 1
+8.   x1 = CMOV(x1, c2, e1)   // if (t1 + t2) == 0, set x1 = -1 / Z
+7.   x1 = x1 * c1      // x1 = (-B / A) * (1 + (1 / (Z^2 * u^4 + Z * u^2)))
+8.  gx1 = x1^2
+9.  gx1 = gx1 + A
+10. gx1 = gx1 * x1
+11. gx1 = gx1 + B            // gx1 = g(x1) = x1^3 + A * x1 + B
+12.  x2 = t1 * x1            // x2 = Z * u^2 * x1
+13.  t3 = gx1^((p + 1) / 4)  // if gx1 is square, this is sqrt(g(x1))
+14.  t4 = t3 * c3
+15.  t4 = t4 * u^3           // if gx1 is not square, this is sqrt(g(x2))
+16.  e3 = t3^2 == gx1
+17.   x = CMOV(x2, x1, e3)   // if e2 == True, x = x1, else x = x2
+18.   y = CMOV(t4, t3, e3)   // if e2 == True, y = t3, else y = t4
+19.  e4 = sgn0(u) == sgn0(y)
+20.   y = CMOV(-y, y, e4)
+21. Output h * (x, y)
 ~~~
 
-### Simplified SWU Method {#simple-swu}
+## Encodings for Montgomery curves
+
+### Elligator 2 Method {#elligator2}
 
-The map2curve_simple_swu(alpha) implements a simplified version of
-Shallue-Woestijne-Ulas algorithm given by Brier et al. {{BR01}}.
+The function map2curve\_elligator2(alpha) implements Elligator 2 {{BHKL13}} for
+curves defined by y^2 = x^3 + A * x^2 + B * x such that A * B * (A^2 - 4 * B) != 0
+and A^2 - 4 * B is non-square in F.
 
-Preconditions: A Weierstrass curve over F such that A!=0, B!=0, and p=3 (mod 4).
+Preconditions: A Montgomery curve where A != 0, B != 0, and A^2 - 4 is non-square in F.
 
 Input: alpha, an octet string to be hashed.
 
-Constants: A and B, the parameters of the Weierstrass curve.
+Constants:
+
+- A and B, the parameters of the curve
+
+- Z, the smallest (in absolute value) non-square in F, breaking ties by choosing
+  the positive value.
+
+Output: (x, y), a point on E.
 
-Output: (x,y), a point on E.
+Sign of y: for u = hash2base(alpha), this encoding ignores the sign of u.
+Thus, we set sgn0(y) == sgn0(u).
+
+Exceptions: The exceptional case is Z * u^2 == -1, i.e., 1 + Z * u^2 == 0.
+Implementations must detect this case and set x1 = -A.
+Note that this can only happen when q = 3 (mod 4).
 
 Operations:
 
 ~~~
 1.   u = hash2base(alpha)
-2.  x1 = (-B / A) * (1 + (1 / (u^4 - u^2)))
-3. gx1 = x1^3 + A*x1 + B
-4.  x2 = -u^2 * x1
-5. gx2 = x2^3 + A*x2 + B
-6. If gx1 is square, set x = x1 and y = sqrt(gx1)
-7. If gx2 is square, set x = x2 and y = sqrt(gx2)
-8. Output h*(x, y)
+2.  x1 = -A * inv0(1 + Z * u^2)
+3.  If x1 == 0, set x1 = -A.
+4. gx1 = x1^3 + A * x1^2 + B * x1
+5.  x2 = -x1 - A
+6. gx2 = x2^3 + A * x2^2 + B * x2
+8.  If is_square(gx1), set x = x1 and y = sqrt(gx1)
+9.  If is_square(gx2), set x = x2 and y = sqrt(gx2)
+10. If sgn0(u) != sgn0(y), set y = -y
+11. Output h * (x, y)
 ~~~
 
-#### Implementation
+#### Implementation, q=3 (mod 4)
 
-The following procedure implements the Simple SWU's algorithm in a straight-line
-fashion.
+The following procedure implements Elligator 2 in a straight-line
+fashion for curves where q=3 (mod 4), including Curve448.
```

Comment:
Sorry, that's not quite correct.

The reason is that the implementations described here are optimized in a way that makes non-black-box use of the sqrt function. See lines 13--16 of this algorithm description, and compare with lines 10--19 of the algorithm for p = 5 mod 8. This optimization is described in the 3rd paragraph of Section 5.5 of the Elligator paper.

I think there's a higher-level issue to be discussed here, which I also mentioned in a response above: does this document want to be a general guide for these algorithms, or does it want to give specific implementation advice for curves of interest? If the latter, then these should be broken out and treated separately, because sqrt() is not just a black box here.

By the way, the same is true for Simplified SWU: one can avoid computing a second square root with non-black-box use of sqrt().

One final thought: if this document ends up describing algorithms for projective coordinates (which are better from an efficiency perspective), then there will be even more non-black-box use of sqrt to avoid taking multiplicative inverses.

From my perspective, giving implementors the best possible implementations will make this document most useful to them. But of course there are plenty of arguments in both directions.

Reply:
One can make the distinction between algorithm and implementation.
As you can see, in the document there is a short description using black-box functions and non-constant time pseudo-code, and then, it is followed by a constant-time implementation. This short description must be enough to be used by anyone on the specified curve.

Also I am aware that by unboxing functions and working on projective coordinates allows more optimization tricks, and definitely, this is more useful for implementors. However, we need to figure out in which part of the draft to include the optimized versions.

Reply:
I agree about the distinction between algorithm and implementation. The *algorithm*, a few lines above, is fully black-box. The two *implementations* (for p = 3 mod 4 and p = 5 mod 8) make non-black-box use. So we seem to be in agreement.

I agree with you that we need to make a decision about optimized versions. I created a new issue, #111, for this. I'll write a specific proposal there in the near future.

Reply:
Good work.